### PR TITLE
1334/activity-orders-from-api

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Activity orders from API: "Your orders, wherever you are"

Main changes:
- At start/network/account change, fetches orders from API
- Shows only orders for connected account/network
- Shows only txs for connected account/network
- Displays at most 10 regulars orders, and as many pending as there are
- Shows orders sorted by creation date, descending
- Replaced `Clear activity` with `View all orders` ~(Pending CSS work)~
- Added a `View all orders` at the bottom of activity list

# To Test

1. Connect a wallet with existing orders
2. Open Account Activities sidebar
* There should be 10 orders or less, doesn't matter how old they are (There might be more if you have more than 10 pending orders)
* There should be a link on top right that takes you to the Explorer to see all of your orders
screenshot_2021-09-27_16-24-30

3. Change account AND/OR network
* Orders should be updated to match the selected account/network
4. Place an order
5. While it's pending, open the same page in another browser / device (:warning: important to test in a different browser otherwise the same local storage will be used)
* The pending order should be visible as pending in the new browser / device

# *NOT* part of this PR

- Load txs the same way as orders
- Refresh pending orders in the background - right now they are updated only on account/network change
- Show orders that have Tokens not yet loaded in the current UI

### CODE REVIEW

**Comment by elena-zh** [comment]
Hey @alfetopito , 
in this pr I see loaded orders in the console, not in the side bar. Is this OK? 
image


My question is: in the side bar we store not only Buy/Sell order logs, but Approve/War/Unwrap transactions logs. Will these logs be loaded from API and displayed in the Side bar on each device where I connect my wallet?

**Comment by alfetopito** [comment]
> Hey @alfetopito ,
> in this pr I see loaded orders in the console, not in the side bar. Is this OK?
> image
> 
> My question is: in the side bar we store not only Buy/Sell order logs, but Approve/War/Unwrap transactions logs. Will these logs be loaded from API and displayed in the Side bar on each device where I connect my wallet?

Indeed, transaction history is not yet loaded. It's not supposed to be part of this change yet.
Also, the version you tested was missing a few changes. I just fixed the build and it should have what I mentioned yesterday.

I'll still have to finalize this PR though :)

File: src/custom/state/orders/reducer.ts [hunk 1]
```diff
@@ -108,6 +109,59 @@ export default createReducer(initialState, (builder) =>
       delete state[chainId].expired[id]
       delete state[chainId].cancelled[id]
     })
+    .addCase(addOrUpdateOrdersBatch, (state, action) => {
```
Comment (alfetopito) [inline src/custom/state/orders/reducer.ts:112]: This might be over-complicated. Suggestions to improve?

**Review by elena-zh — COMMENTED** [review]
HI @alfetopito , awesome!
Some minor issues: 
1. There is not 'Cancelling' state in the 2nd device/tab/browser when cancel an order in the 1st one: the order remains to be in 'Open' status instead of 'Cancelling'
cancelling

2. Can't understand why, but [1 order](https://protocol-explorer.dev.gnosisdev.com/rinkeby/orders/0x2b3f76bb5ac25a4f0c752c0f1ec972866941781f77dd0600305c2fb7e54a8b9dff714b8b0e2700303ec912bd40496c3997ceea2b6152f7df) is always skipped from the activity sidebar list for 1 my account in Rinkeby.
1 order is missing

3. You mentioned, that CSS work is pending, however, I'd mention it here: bar with action button is shifted when scroll the side bar to the bottom (related to this pr only): 
recent shifted

4. As an enhancement (also related to CSS), it would be nice to display 'View all orders' at the bottom of the list: it is super user-friendly, especially on mobile devices
at the bottom

File: src/custom/state/orders/updaters/APIOrdersUpdater.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  }
+}
+
+export function APIOrdersUpdater(): null {
+  const { account, chainId } = useActiveWeb3React()
+  const allTokens = useAllTokens()
+  const addOrUpdateOrdersBatch = useAddOrUpdateOrdersBatch()
+
+  useEffect(() => {
+    if (account && chainId && Object.keys(allTokens).length > 0) {
```
Comment (W3stside) [inline src/custom/state/orders/updaters/APIOrdersUpdater.ts:43]: i would cache the object keys into a const for readability

File: src/custom/state/orders/updaters/APIOrdersUpdater.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+      getOrders(chainId, account, 100)
+        .then((_orders) => {
+          console.log(`APIOrdersUpdater::Fetched ${_orders.length} orders for account ${account} on chain ${chainId}`)
+
+          // Transform API orders into internal order objects
+          const orders = _orders.reduce<Order[]>((acc, order) => {
```
Comment (W3stside) [inline src/custom/state/orders/updaters/APIOrdersUpdater.ts:49]: can this be extrapolated into a helper function outside the function body?

**Comment by alfetopito** [comment]
> HI @alfetopito , awesome! Some minor issues:
> 
>     1. There is not 'Cancelling' state in the 2nd device/tab/browser when cancel an order in the 1st one: the order remains to be in 'Open' status instead of 'Cancelling'
>        cancelling

The cancelling state is local and not part of the API.
I'll see what can be done, but at first I don't think this will be sync-able.
 
>     2. Can't understand why, but [1 order](https://protocol-explorer.dev.gnosisdev.com/rinkeby/orders/0x2b3f76bb5ac25a4f0c752c0f1ec972866941781f77dd0600305c2fb7e54a8b9dff714b8b0e2700303ec912bd40496c3997ceea2b6152f7df) is always skipped from the activity sidebar list for 1 my account in Rinkeby.
>        1 order is missing

I suspect you found a TODO :sweat_smile: 

Can you check in your console logs whether you see a warning starting with `APIOrdersUpdater::Tokens not found for order`

In summary, one of the tokens is not loaded, thus the order is ignored.
 
>     3. You mentioned, that CSS work is pending, however, I'd mention it here: bar with action button is shifted when scroll the side bar to the bottom (related to this pr only):
>        recent shifted

Weird. Will try to fix it.
 
>     4. As an enhancement (also related to CSS), it would be nice to display 'View all orders' at the bottom of the list: it is super user-friendly, especially on mobile devices
>        at the bottom

That's a good idea! Will work on it.

**Comment by alfetopito** [comment]
@elena-zh 

Addressed `1` and `4`

Working on `2`, might come as another PR as it won't be trivial

`3` is already in prod, and I don't know how to fix it. @biocom maybe?

**Comment by elena-zh** [comment]
> Can you check in your console logs whether you see a warning starting with APIOrdersUpdater::Tokens not found for order
> In summary, one of the tokens is not loaded, thus the order is ignored.

No, I do not see any error from my side. Moreover, I can find this order in the BE request, but it is not displayed on the UI
image

**Comment by elena-zh** [comment]
> 3 is already in prod, and I don't know how to fix it. @biocom maybe?

Hm. I have not encountered this issue neither in Prod, nor in Dev. 
prod
dev

May be just kind of transparency issue
image


But anyways, I have created a separate issue for this case #1489

**Comment by biocom** [comment]
@elena-zh What browser do you experience the non transparent issue?

**Comment by elena-zh** [comment]
@biocom , Firefox. I mentioned in in #1489

**Comment by biocom** [comment]
One issue I see is the 'VIEW ALL ORDERS' at the bottom. In my view, this is unnecessary given it's already shown in the sticky top bar. But if we want to keep it, the <a href> should be a child of the same parent DIV as the order 'cards'. Otherwise it messes up the blue X sticky bar on top.

**Comment by biocom** [comment]
As for the transparency issue, I think this is because Firefox lacks support of the backdrop-filter property:
https://developer.mozilla.org/en-US/docs/Web/CSS/backdrop-filter

There's ways to overcome this (@support CSS testing and fallback for Firefox using a SVG blur effect). But would address this in a separate PR.

**Comment by elena-zh** [comment]
@alfetopito , 
As for the 'Cancelled' orders sync, everything works fine when open the app in different browsers/tabs in WEB.
However, I still face issues when I cancel an order in a desktop browser (Chrome), then connect my wallet in a mobile device and cancel an order from there (or vice versa). 
See videos: 
1. Order is cancelled in a mobile device, I open the app in a desktop browser: https://watch.screencastify.com/v/jiiTBJ8zDnIhuPbfJvi1
https://watch.screencastify.com/v/pAD31olFdLPStcmmuRiV

2. Order is cancelled in a desktop browser, I open the app in a mobile device (does not matter Android or iOS): 

https://user-images.githubusercontent.com/70885163/135277885-c62543a3-9e86-42a6-a6b2-92fb99b019f4.mp4

If it needed, I can create a separate issue for these cases...

File: src/custom/state/orders/reducer.ts (hunk 1 above)
Comment (W3stside) [reply src/custom/state/orders/reducer.ts:112]: i wouldn't do all of the "business logic" inside the reducer, may be better to split this into logic elsewhere no?

**Comment by alfetopito** [comment]
> @alfetopito , As for the 'Cancelled' orders sync, everything works fine when open the app in different browsers/tabs in WEB. However, I still face issues when I cancel an order in a desktop browser (Chrome), then connect my wallet in a mobile device and cancel an order from there (or vice versa). See videos:
> 
>     1. Order is cancelled in a mobile device, I open the app in a desktop browser: https://watch.screencastify.com/v/jiiTBJ8zDnIhuPbfJvi1
>        https://watch.screencastify.com/v/pAD31olFdLPStcmmuRiV
> 
>     2. Order is cancelled in a desktop browser, I open the app in a mobile device (does not matter Android or iOS):
> 
> 
> sync.mp4
> 
> If it needed, I can create a separate issue for these cases...

You were 100% correct Elena. There was a bug where the `cancelling` status wasn't really syncing.
It worked on another tab because it's the same browser thus using the same local storage.
Fixed now :crossed_fingers: 

----

> One issue I see is the 'VIEW ALL ORDERS' at the bottom. In my view, this is unnecessary given it's already shown in the sticky top bar. But if we want to keep it, the should be a child of the same parent DIV as the order 'cards'. Otherwise it messes up the blue X sticky bar on top.

@biocom I don't really see the sticky header when I reach the bottom. About half way through it slides away. 
That's the same on Firefox and Chrome on desktop and also on MM mobile browser:
Screenshot_20210929-150353

**Comment by biocom** [comment]
@alfetopito This is what I mean (the top X sticky bar disappearing):

https://user-images.githubusercontent.com/31534717/135413896-3283e2b2-9c91-443b-a05d-94442739c366.mov

**Comment by elena-zh** [comment]
> You were 100% correct Elena. There was a bug where the cancelling status wasn't really syncing.
> It worked on another tab because it's the same browser thus using the same local storage.
> Fixed now 🤞

I have verified: works fine now. Thanks!

**Comment by alfetopito** [comment]
> @alfetopito This is what I mean (the top X sticky bar disappearing):
> Screen.Recording.2021-09-30.at.10.11.22.mov

Yep, I see that too, also on Prod on both mobile and desktop (FF and Chrome)

**Review by anxolin — COMMENTED** [review]
Awesome! 

I think this feature deserves a release, although will be blocked by the explorer (i guess this will go out already soon anyways).

Mostly nitpicks. Code works as a charm, the only concerns I have are:
- Is it possible this updater mutate an order that we are waiting on, so we don't notify when it's filled or when some important change of data happen? (in other words, can this interfere and be a race condition to other workflows?) 
- Can the updater logic be making more queries than needed if we keep adding tokens? or this is not a concerning? One alternative (not sure if its good proposal), u make the `useEffect` of the updater to just set the one `useState` with the raw orders. Then you hafe a different `useEffect` that on any change in `allOrders` or in `allTokens`, it would recompute the list of orders and update them.

File: src/custom/hooks/useRecentActivity.ts [hunk 4]
```diff
@@ -31,69 +31,62 @@ enum TxReceiptStatus {
...
+          // and time in MS vs ISO string as Orders uses
+          ({
+            ...order,
+            addedTime: Date.parse(order.creationTime),
+          })
+        )
+        // sort orders by calculated `addedTime` descending
+        .sort((a, b) => b.addedTime - a.addedTime)
+        // show at most 10 regular orders, and as much pending as there are
+        .filter((order, index) => index < 10 || order.status === OrderStatus.PENDING)
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/custom/hooks/useRecentActivity.ts:62]: Magic number fixed.

One thing to note is that, we limit to at most 10 non-pending orders, but show all pending orders that exist (that's why I fetch 100 orders).

Regarding the forever storing all orders, that's already the case right now. Things older than 24h were filtered out, but they were still there :)

What about cleaning orders older than 1 week and only from the 100 oldest order and above?
Could set it to run once a day/once per page load

File: src/custom/state/orders/reducer.ts [hunk 5]
```diff
@@ -108,6 +109,66 @@ export default createReducer(initialState, (builder) =>
...
+        } else if (fulfilled[id]) {
+          orderObj = fulfilled[id]
+          delete fulfilled[id]
+        } else if (expired[id]) {
+          orderObj = expired[id]
+          delete expired[id]
+        } else if (cancelled[id]) {
+          orderObj = cancelled[id]
+          delete cancelled[id]
+        }
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/custom/state/orders/reducer.ts:138]: Let's see who wins the merge race :D

And yes, I've seen your refactors and would like to use them here

File: src/custom/state/orders/reducer.ts [hunk 6]
```diff
@@ -108,6 +109,66 @@ export default createReducer(initialState, (builder) =>
...
+            cancelled[id] = { order, id }
+            break
+          case 'expired':
+            expired[id] = { order, id }
+            break
+          case 'fulfilled':
+            fulfilled[id] = { order, id }
+            break
+          default:
+            // TODO: add it regardless?
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/custom/state/orders/reducer.ts:167]: This made more sense for the way status was calculated before. Anyway, I've refactored and will log an error if the status is even empty

File: src/custom/state/orders/updaters/APIOrdersUpdater.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+export function APIOrdersUpdater(): null {
+  const { account, chainId } = useActiveWeb3React()
+  const allTokens = useAllTokens()
+  const tokenAreLoaded = useMemo(() => Object.keys(allTokens).length > 0, [allTokens])
+  const addOrUpdateOrdersBatch = useAddOrUpdateOrdersBatch()
+
+  useEffect(() => {
+    if (account && chainId && tokenAreLoaded) {
+      getOrders(chainId, account, 100)
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/custom/state/orders/updaters/APIOrdersUpdater.ts:93]: I've explained on https://github.com/gnosis/cowswap/pull/1395//#discussion_r722421384

Short version: there's no limit for pending orders, so I load far enough that should account for all pending orders

When there's such a filter on the backend this can be revisited.

And I'll remove the magic number

File: src/index.tsx [hunk 8]
```diff
@@ -71,6 +76,7 @@ function Updaters() {
       <CancelledOrdersUpdater />
       <FeesUpdater />
       <UnfillableOrdersUpdater />
+      <APIOrdersUpdater />
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/index.tsx:79]: Because there are 3 other OrdersUpdater already:
screenshot_2021-10-05_11-01-56

Having one that's just `OrdersUpdaters` would be ambiguous IMO

File: src/custom/state/orders/updaters/APIOrdersUpdater.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+          }, [])
+
+          // Add to redux state
+          addOrUpdateOrdersBatch({ orders, chainId })
+        })
+        .catch((e) => {
+          console.error(`APIOrdersUpdater::Failed to fetch orders`, e)
+        })
+    }
+  }, [account, addOrUpdateOrdersBatch, allTokens, chainId, tokenAreLoaded])
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/custom/state/orders/updaters/APIOrdersUpdater.ts:113]: I wouldn't say intended but it's a side effect.
Trying to avoid as much as possible to fetch tokens from the network.

Any suggestion to better handle this?

File: src/custom/state/orders/updaters/APIOrdersUpdater.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+          const orders = apiOrders.reduce<Order[]>((acc, order) => {
+            const storeOrder = transformApiOrderToStoreOrder(order, chainId, allTokens)
+            if (storeOrder) {
+              acc.push(storeOrder)
+            }
+            return acc
+          }, [])
+
+          // Add to redux state
+          addOrUpdateOrdersBatch({ orders, chainId })
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/custom/state/orders/updaters/APIOrdersUpdater.ts:107]: `Upsert` sounds wrong, I'll remove the `batch` part.

File: src/custom/state/orders/updaters/APIOrdersUpdater.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/custom/state/orders/updaters/APIOrdersUpdater.ts:113]: Just saw the suggestion on general comment.

I'll give that a try, but not on this PR. I'll do it either on https://github.com/gnosis/cowswap/pull/1504 or as follow up to it.

**Review by elena-zh — APPROVED** [review]
LGTM!
Issues with missing log entries are fixed in #1504 and #1536

File: src/custom/api/gnosisProtocol/api.ts [hunk 11]
```diff
@@ -376,5 +398,14 @@ export async function getGasPrices(chainId: ChainId = DEFAULT_NETWORK_FOR_LISTS)
 
 // Register some globals for convenience
 registerOnWindow({
-  operator: { getFeeQuote, getAppDataDoc, getOrder, sendSignedOrder, uploadAppDataDoc, apiGet: _get, apiPost: _post },
+  operator: {
```
Comment (anxolin) [inline src/custom/api/gnosisProtocol/api.ts:401]: nit: We don't call it operator anymore :) 

gpApi?

File: src/custom/components/AccountDetails/index.tsx [hunk 12]
```diff
@@ -148,17 +145,8 @@ export default function AccountDetails({
...
-  const clearAllActivityCallback = useCallback(() => {
-    if (chainId) {
-      batch(() => {
-        dispatch(clearAllTransactions({ chainId }))
-        dispatch(clearOrders({ chainId }))
-      })
-    }
-  }, [dispatch, chainId])
+
+  const explorerOrdersLink = account && connectedChainId && getExplorerAddressLink(connectedChainId, account)
```
Comment (anxolin) [inline src/custom/components/AccountDetails/index.tsx:149]: I fear some merge conflicts with out presign stuff

I'm not sure what will be best to merge first...

**Review by anxolin — APPROVED** [review]
Nice! It's a pity all my comments were lost, so I kind of went to the email to see some of the replies, but my question/comment was not there. I guessed what my question were by reading your answers, but it was hard to follow up

I wish github would keep those even after a rebase! 


Anyways, looks solid!

File: src/custom/components/AccountDetails/index.tsx (hunk 12 above)
Comment (alfetopito) [reply src/custom/components/AccountDetails/index.tsx:149]: Oh yeah, there will be conflicts.

I'll merge first =P

After all you need this for your feature

File: src/index.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (anxolin) [reply src/index.tsx]: 🙈

File: src/custom/components/AccountDetails/index.tsx [hunk 13]
```diff
@@ -210,7 +198,7 @@ export default function AccountDetails({
               {chainId && account && (
                 <AddressLink
                   hasENS={!!ENSName}
-                  isENS={ENSName ? true : false}
+                  isENS={!!ENSName}
```
Comment (anxolin) [inline src/custom/components/AccountDetails/index.tsx:201]: nitpick: should we create a const `isENS` just to keep it dry-er

**Review by anxolin — APPROVED** [review]
